packages/hydrooj/src/handler/manage.ts

```diff
@@ -319,6 +324,79 @@ class SystemUserPrivHandler extends SystemHandler {
     }
 }
 
+class SystemBulkRejudgeHandler extends SystemHandler {
+    @requireSudo
+    async get() {
+        this.response.template = 'manage_bulk_rejudge.html';
+        this.response.body = {
+            filterPid: null,
+            filterTid: null,
+            filterUidOrName: null,
+            filterLang: null,
+            filterStatus: null,
+            message: '',
+        };
+    }
+
+    @requireSudo
+    @param('pid', Types.ProblemId, true)
+    @param('tid', Types.ObjectId, true)
+    @param('uidOrName', Types.UidOrName, true)
+    @param('lang', Types.String, true)
+    @param('status', Types.Int, true)
+    @param('draft', Types.Boolean)
+    async post(domainId: string, pid: string, tid: ObjectId, uidOrName: string, lang: string, status: number, draft = false) {
+        if (!pid && !tid && !uidOrName && !lang && status === undefined) throw new ValidationError();
+        const q: Filter<RecordDoc> = { 'files.hack': { $exists: false } };
+        let tdoc = null;
+        if (uidOrName) {
+            const udoc = await user.getById(domainId, +uidOrName)
+                || await user.getByUname(domainId, uidOrName)
+                || await user.getByEmail(domainId, uidOrName);
+            if (udoc) q.uid = udoc._id;
+            else throw new UserNotFoundError(uidOrName);
+        }
+        if (tid) {
+            if ([record.RECORD_GENERATE, record.RECORD_PRETEST].includes(tid)) throw new ValidationError('tid');
+            tdoc = await contest.get(domainId, tid);
+            if (tdoc) q.contest = tid;
+            else throw new ContestNotFoundError(domainId, tid);
+        } else q.contest = { $nin: [record.RECORD_GENERATE, record.RECORD_PRETEST] };
+        if (pid) {
+            if (typeof pid === 'string' && tdoc && /^[A-Z]$/.test(pid)) {
+                pid = tdoc.pids[parseInt(pid, 36) - 10];
+            }
+            const pdoc = await problem.get(domainId, pid);
+            if (pdoc) q.pid = pdoc.docId;
+            else throw new ProblemNotFoundError(domainId, pid);
+            if (!pdoc.config || typeof pdoc.config === 'string') throw new ProblemConfigError();
+        }
+        if (lang) q.lang = lang;
+        if (typeof status === 'number') q.status = status;
+        this.response.body = {
+            filterPid: pid,
+            filterTid: tid,
+            filterUidOrName: uidOrName,
+            filterLang: lang,
+            filterStatus: status,
+        };
+        if (draft) {
+            const count = await record.coll.count(q);
+            this.response.body.message = `${count} records found, please confirm.`;
+            this.response.template = 'manage_bulk_rejudge.html';
+        } else {
+            const rdocs = await record.getMulti(domainId, q).project({ _id: 1, contest: 1 }).toArray();
```

Comment:
this can be an ultra slow query taking several minutes on large collection, converting into background task would be preferred.

Reply:
Convert request to task?